kv/util/runtime/options.go

```diff
@@ -0,0 +1,133 @@
+// Copyright (c) 2017 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package runtime
+
+import (
+	"time"
+
+	"github.com/m3db/m3cluster/kv"
+	"github.com/m3db/m3cluster/kv/mem"
+	"github.com/m3db/m3x/instrument"
+)
+
+const (
+	defaultInitWatchTimeout = 10 * time.Second
+)
+
+// Options provide a set of value options.
+type Options interface {
+	// SetInstrumentOptions sets the instrument options.
+	SetInstrumentOptions(value instrument.Options) Options
+
+	// InstrumentOptions returns the instrument options.
+	InstrumentOptions() instrument.Options
+
+	// SetInitWatchTimeout sets the initial watch timeout.
+	SetInitWatchTimeout(value time.Duration) Options
+
+	// InitWatchTimeout returns the initial watch timeout.
+	InitWatchTimeout() time.Duration
+
+	// SetKVStore sets the kv store.
+	SetKVStore(value kv.Store) Options
+
+	// KVStore returns the kv store.
+	KVStore() kv.Store
+
+	// SetUnmarshalFn sets the unmarshal function.
+	SetUnmarshalFn(value UnmarshalFn) Options
+
+	// UnmarshalFn returns the unmarshal function.
+	UnmarshalFn() UnmarshalFn
+
+	// SetProcessFn sets the process function.
+	SetProcessFn(value ProcessFn) Options
+
+	// ProcessFn returns the process function.
+	ProcessFn() ProcessFn
+}
+
+type options struct {
+	instrumentOpts   instrument.Options
+	initWatchTimeout time.Duration
+	kvStore          kv.Store
+	unmarshalFn      UnmarshalFn
+	processFn        ProcessFn
+}
+
+// NewOptions creates a new set of options.
+func NewOptions() Options {
+	return &options{
+		instrumentOpts:   instrument.NewOptions(),
+		initWatchTimeout: defaultInitWatchTimeout,
+		kvStore:          mem.NewStore(),
```

Comment:
Hm using the memstore feels like a reasonable default to me, leaving it nil is slightly better but we are back to the same place if the users don't call `Validate`.